### ISSUE TITLE
same devices have same vidpid's

### ISSUE THREAD

**Issue by josephlarralde** [Lucarda/pd-hidraw#5]
Hi, thanks for creating and maintaining this external that is very useful to me these days.

I see a limitation that could be overcome, though :
I'd like to be able to connect devices by some kind of unique hardware id regardless of how they are plugged to the computer, and I have some logitech joysticks of the same model that all have the same vidpid, so I can't use the open-vidpid message to connect a specific device.

I noticed in the console that the only thing that distinguishes them is the "path: DevSrvsID:xxxxxxxxxx" line.
At the moment I don't have enough spare time to dig into this myself, so I was wondering if it would be possible to have an "open-path" message that would take the "DevSrvsID:xxxxxxxxxx" string in the same way the open-vidpid message works.

An alternative solution could be to send the devicelist details that are printed to the console as a bunch of messages that could be parsed from within pd to retrieve the correct device enums, as hacking the pd console stream seems a bit convoluted ...

**Comment by josephlarralde** [comment]
I'll try to take a look at this whenever it becomes critical to me ... cheers !

EDIT : it seems that the path is not consistent across device (dis)connections, and all the serial numbers I get are empty strings. So implementing an open-path message would be useless, and no more luck with serial numbers ... :/
I'll look into the alternative solution(s)

**Comment by josephlarralde** [comment]
Ok, I see that I can open several devices with the same vidpid's one after another with different [hidraw] instances.
I think I will just setup a simple device identification workflow in my patches and give up on this since I see no obvious elegant solution.

Sorry for the noise, feel free to close the issue !

**Comment by Lucarda** [comment]
Hi @josephlarralde

devices opened via [listdevices( and [open $1( are opened via path but this is done internally in the object. one of the reasons there's no such method in the object is that on Windows paths contain forbidden characters (in Pd messages) as "\\" , "{" and "}".

i'm now thinking if an extra outlet with some of the fields in [listdevices( could be added.

**Comment by Lucarda** [comment]
may be a simple list with with 3 items for the 3th outlet would be `<float: enum>` `<symbol: Manufacturer>` `<symbol: Product>`:

```
print: 1 (Standard\ system\ devices) Generic\ USB\ K/B
print: 2 foo bar
print: ...
```
i'm not sure how useful this could be as there are duplicates of the same single device when doing [listdevices(

still thinking of it.

**Comment by Lucarda** [comment]
also note that VID is vendor Id and PID is product Id so two same model HIDs have (of course) same vidpid.

:no_good:

**Comment by josephlarralde** [comment]
Hi @Lucarda thanks for answering ! I think I should have experimented more before opening this issue ... I only tried [hidraw] long enough on MacOS to realize that no device at all gives me a serial number, so I guess this is common practice for manufacturers to not provide any. I feel like this is a dead end because the system and the devices simply don't provide enough useful information to identify a physical device between others that have the same vid / pid.

My plan is now to have a global "assignation" mode that will prepend the right id to all messages coming out from each abstraction containing a [hidraw] after all of them have been open-vidpid'ed (with a loadbang). This should work (and be a light enough procedure for users of the patch I'm developing) because it seems that once a device has been opened, the next open-vidpid message will not open it because it's "busy", and will instead open one that still has not been opened.

I remember running into similar difficulties with [comport] : ports needing to be hardcoded (or saved into preset files) according to the machine the patch is running on, so I doubt we could find an entirely satisfying solution, except if hidapi is not exposing all functionnalities, which would be surprising because, AFAIK, libusb is the main library around allowing to talk to usb devices ...

**Comment by josephlarralde** [comment]
Oooops, once again I talked too fast : in fact there is no "busy state check" at all, it just sometimes works as expected, sometimes not (several hidraw's connect to the same device and other devices are ignored). I'll go back to my fork of your repository and dig to see if I can find a way to fix this. Meanwhile, if you have any advice to offer, please let me know. Cheers !

EDIT : my assignation workflow is working, so I just need a way to connect all devices without having to look at the whole log from the console. Your previous suggestion might be just what I need ... I think I'll try this first.

But you said that sometimes the system reports duplicate devices ? I didn't witness this behaviour yet with gamepads and joysticks but this doesn't mean I will not in the future ... all the more I'm developing on macos but my end users run linux ... Anyway I'll let you know if I get something to work reliably on both platforms.

**Comment by Lucarda** [comment]
about the duplicates i think i only saw them in Windows and for keyboards (which are anyway non accessible in this platform)

i think the new outlet with the 3 items list will do what's needed for this situation and might be handy on others too.

i'll work on it ASAP but feel free to do it in your fork and eventually send a PR, if the thing is simple and straight forward i'll surely merge it. :)

**Comment by Lucarda** [comment]
also note that [hidraw] is also part of purr-data and (i think) pd-l2ork. i might have to be totally sure of what to do and not to break backwards compatibility and all that. but of course i count on you on this decision makings :)

**Comment by Lucarda** [comment]
@josephlarralde can you test branch `squash012` from:

https://github.com/Lucarda/pd-hidraw/tree/squash012

seems fine here thought only tested under Windows.

**Comment by josephlarralde** [comment]
Testing right now (nothing plugged in, only with Apple integrated hardware) : I witness a strange behaviour where I don't get the same results each time I press the listdevices messages ... still investigating but we're mostly there, thanks for the quick update !

**Comment by Lucarda** [comment]
>  I witness a strange behaviour where I don't get the same results each time I press the listdevices messages ... 

strange. please investigate as i don't have a mac.

I had just did a fast check on Debian12 and it seems fine (only doing listdevices).

**Comment by Lucarda** [comment]
this is what i get on Debian12 on 2 consecutive [listdevices(. it shows what i mention about duplicate (or triplicated) devices. it's a live usb stick with an usbhub with 1 mouse and 1 keyboard. nothing else is attached on the pc box.

```
print: 1 SEM USB\ Keyboard
-----------
Pd device enum: 1
device VID PID (shown in decimal notation): 6700 16932


  type: 1a2c 4224
  path: /dev/hidraw0
  serial_number: 
  Manufacturer: SEM
  Product:      USB Keyboard
  Release:      110
  Interface:    0
  Usage (page): 0x6 (0x1)
 
print: 2 SEM USB\ Keyboard
-----------
Pd device enum: 2
device VID PID (shown in decimal notation): 6700 16932


  type: 1a2c 4224
  path: /dev/hidraw1
  serial_number: 
  Manufacturer: SEM
  Product:      USB Keyboard
  Release:      110
  Interface:    1
  Usage (page): 0x1 (0xc)
 
print: 3 SEM USB\ Keyboard
-----------
Pd device enum: 3
device VID PID (shown in decimal notation): 6700 16932


  type: 1a2c 4224
  path: /dev/hidraw1
  serial_number: 
  Manufacturer: SEM
  Product:      USB Keyboard
  Release:      110
  Interface:    1
  Usage (page): 0x80 (0x1)
 
print: 4 SIGMACHIP Usb\ Mouse
-----------
Pd device enum: 4
device VID PID (shown in decimal notation): 7247 3


  type: 1c4f 0003
  path: /dev/hidraw2
  serial_number: 
  Manufacturer: SIGMACHIP
  Product:      Usb Mouse
  Release:      110
  Interface:    0
  Usage (page): 0x2 (0x1)
 

```

second run:

```
print: 1 SEM USB\ Keyboard
-----------
Pd device enum: 1
device VID PID (shown in decimal notation): 6700 16932


  type: 1a2c 4224
  path: /dev/hidraw0
  serial_number: 
  Manufacturer: SEM
  Product:      USB Keyboard
  Release:      110
  Interface:    0
  Usage (page): 0x6 (0x1)
 
print: 2 SEM USB\ Keyboard
-----------
Pd device enum: 2
device VID PID (shown in decimal notation): 6700 16932


  type: 1a2c 4224
  path: /dev/hidraw1
  serial_number: 
  Manufacturer: SEM
  Product:      USB Keyboard
  Release:      110
  Interface:    1
  Usage (page): 0x1 (0xc)
 
print: 3 SEM USB\ Keyboard
-----------
Pd device enum: 3
device VID PID (shown in decimal notation): 6700 16932


  type: 1a2c 4224
  path: /dev/hidraw1
  serial_number: 
  Manufacturer: SEM
  Product:      USB Keyboard
  Release:      110
  Interface:    1
  Usage (page): 0x80 (0x1)
 
print: 4 SIGMACHIP Usb\ Mouse
-----------
Pd device enum: 4
device VID PID (shown in decimal notation): 7247 3


  type: 1c4f 0003
  path: /dev/hidraw2
  serial_number: 
  Manufacturer: SIGMACHIP
  Product:      Usb Mouse
  Release:      110
  Interface:    0
  Usage (page): 0x2 (0x1)
```

**Comment by josephlarralde** [comment]
Oh wait, it's just me confused about the ton of Apple devices with many weird names and duplicates, and their order changing each time I hit [listdevices<, which appears to be normal, and the list is always the same in a different order.
I hadn't noticed before because I was not hitting [listdevices< that often.
Also I'm at home so I was trying to connect my old-but-not-too-old logitech F310 gamepad for the first time, but it seems apple doesn't support some logitech devices anymore, including this one.
The object works as expected with an older logitech joystick I also have at home (a logitech attack3).
I will confirm soon that it works with the setup I have at work, but so far it looks fine to me.
I already have my automatic connection system ready for tomorrow, based on a bunch of [text] objects :)
Sorry for the confusion, and thanks again !

**Comment by josephlarralde** [comment]
Hmmm ... still stuck on my use case. Just tested with 2 Thrustmaster dual analog 3.2 gamepads : besides the fact that for an unknown reason they now both have duplicate entries in the device list (but with different Usage (page) values), I realize that I just can't use a "master" hidraw to list all devices and open them from other hidraw objects without calling listdevices on these other objects, which gives me the same list but each time in a different order, and since there is still no way to identify a specific hardware device e.g. by serial number, I can't know which ones I should open. I'm considering starting to bang my head on the wall ... maybe I should just give up with apple and see if it works better on linux :(

Is your device list scrambled each time you call listdevices ?

I feel Apple is a bit hostile towards gamers anyway, and maybe it's pointless to try fighting their unilateral, IMO  very bad (but probably very "secure") design decisions

**Comment by josephlarralde** [comment]
Ok, I tried [hid] from Hans-Christoph Steiner and it seems I can achieve my goals with it on the versions of macos and linux I'm using, so I will stick with it for now. The codebase is pretty old but it might be worth considering some parts of it as an inspiration for [hidraw] (not sure if it's possible to kind of merge both though, as it doesn't seem to rely on libusb).
Unfortunately I'm sorry to say that I just don't have the time to start digging into this ...
Once again, huge thanks for your reactivity and efforts, I hope I didn't waste too much of your time

**Comment by Lucarda** [comment]
Ok, the new outlet does not bring us any solution so i'll pause it for now. no new release.

closing the issue. feel free to reopen.

:)

**Comment by josephlarralde** [comment]
Yep, this is a good sum up :) 
Thinking out loud : it seems to me that the key feature of hid is that it manages to assign ids to devices consistently. First, I think it sorts devices alphabetically by manufacturer + product name, and it also has a "location" field shown as a hex number that also seems consistent. Not sure if this "location" is equivalent to the "path" in hidraw. I'm still not done implementing the whole multi-device patch and the probability that I run into similar issues is not null yet ... will let you know in this thread, which I might reopen if I have the opportunity to take a deeper dive into both externals.

EDIT : Thinking a little more, I remember (with Max externals, but should work with Pd too) that I managed to share structs across instances of an external, probably simply by defining them as static. This would free us from having to rebuild the devicelist that gets scrambled everytime on macos (13.5 in my case). Maybe that's how hid works, but I didn't read a single line of code from it yet so I can't tell right now.

**Comment by Lucarda** [comment]
> Thinking out loud : it seems to me that the key feature of hid is that it manages to assign ids to devices consistently. First, I think it sorts devices alphabetically by manufacturer + product name

All of these is not done at all in [hidraw]. it's done for us by **hidapi**.


> it also has a "location" field shown as a hex number that also seems consistent. Not sure if this "location" is equivalent to the "path" in hidraw

I think (not looking at [hid]) that this is VID PID. on [hidraw] they are converted to decimal notation on purpose.

**Comment by josephlarralde** [comment]
> All of these is not done at all in [hidraw]. it's done for us by hidapi.

Hmmm I see, so that means hidapi may not show a consistent behaviour on macos, hence the scrambled device list

> I think (not looking at [hid]) that this is VID PID. on [hidraw] they are converted to decimal notation on purpose.

No, vid and pid are also displayed by [hid] as hex numbers but there is no path, that's why I thought they might be the same thing.

But I also notice the working versions of [hid] are not the same for macos and linux and differ slightly in behaviour so I have to use only hex vid, hex pid and open status to get a patch that works on both macos and linux (not tested on windows yet). Kind of ugly because I need to open a fixed guessed number of device ids to grab each one's info (and get a bunch of errors in the console), then I confirm I can open them with same ids from other [hid]'s so it's working for my use case. To be continued ...

Cheers and happy holidays if you get some !

**Comment by ben-wes** [comment]
i've been following this interesting discussion - it would certainly be nice to allow for automated ways of opening multiple devices!

not sure if my thoughts can add to this ... but AFAICT (from what was discussed), the only way of clearly addressing a specific device is through its path (if no serial number is provided). and since the special characters in a path might cause trouble: could it be an option to work with an additional method to connect via a path hash? this could be output on the additional outlet in a list together with PID/VID/serial number/ escaped or sluggified path/manufacturer name/product name/etc.? that way, the scrambled list (and scrambled enumeration) wouldn't matter and a device is still clearly addressable.

also, for this purpose, it might be a nice addition to output/list devices filtered by VID/PID (hidapi provides that option AFAIK)?

**Comment by Lucarda** [comment]
> also, for this purpose, it might be a nice addition to output/list devices filtered by VID/PID (hidapi provides that option AFAIK)?

will look into it.

------------

there is one simple way to output the path in a neutral format: ASCII bytes.

i'll try to append a list of bytes to the 3th outlet: `<float> <symbol> <symbol> <list of bytes>`

then we could have another method [openpath(

[12 54 87 ...(
|
[list append openpath]
|
[hidraw]

?

**Comment by Lucarda** [comment]
>also, for this purpose, it might be a nice addition to output/list devices filtered by VID/PID (hidapi provides that option AFAIK)?

> will look into it.

i see one can enumerate a single VID PID just at the beginning 

`x->devs = hid_enumerate(<VID>, <PID>);`

**Comment by ben-wes** [comment]
> i'll try to append a list of bytes to the 3th outlet: `<float> <symbol> <symbol> <list of bytes>`

sounds good to me, too! - the only disadvantage i see there is that this will rule out the option of possibly adding other symbols or values to the list at a later point if needed?

EDIT: but it's certainly less arbitrary than some hash/token derived from the path string.

**Comment by Lucarda** [comment]
> the only disadvantage i see there is that this will rule out the option of possibly adding other symbols or values to the list at a later point if needed?

this is true. may be we can have an extra object with 4 outlets: 

`<symbol :maunufact> | <symbol :produc> | <list float :VID PID>  | <list float :path>`

say [hidraw/devlist] ?

> also, for this purpose, it might be a nice addition to output/list devices filtered by VID/PID (hidapi provides that option AFAIK)?

true the we can have a different behavior just for this case and add work so that we can have a new method and do:

![hidraw1](https://github.com/user-attachments/assets/bf35e3e7-e6cf-4ac8-aecd-20723d37f379)


not sure if we get the same order of devices in both `[listvidpid 1204 123(` on macOS.

i could have a test code ASAP.

---------------------

anyway would have to think if we need one or the other or both.

**Comment by Lucarda** [comment]
@josephlarralde can you check `listvidpid` branch:

https://github.com/Lucarda/pd-hidraw/tree/listvidpid

we had to thank @ben-wes for this as it may had saved our week :)

doing `[list-vidpid VID  PID(` should enumerate same vidpid devices. I had only tested on Windows.

**Comment by ben-wes** [comment]
> we had to thank @ben-wes for this as it may had saved our week :)

don't thank me too early! - it's quite likely that the enumerations will still get scrambled that way (since it's the same code - just filtered output).

i'll also gladly test soon! unfortunately, i currently don't have 2 devices with similar vid/pid here for testing (but will have on sunday evening).

and i wonder if it should be considered a bug on the hidapi side that the order of that array is not consistent (should be easily fixable on their side)? right now, i'm still afraid that an option to connect via the path will be the only solution for mac - but let's see! thanks for your work and quick reactions here, @Lucarda !

**Comment by Lucarda** [comment]
> don't thank me too early!

:) ok.

i'll try to code `[hidraw/devlist]` so we have some way to:

- use paths 
- have same patch level control of [listdevices(  

i'm not sure the latter is useful.

**Comment by josephlarralde** [comment]
Hey @ben-wes thanks for chiming in !

Thanks @Lucarda for the new features, I'll test everything I can tomorrow at work and report.

> right now, i'm still afraid that an option to connect via the path will be the only solution for mac

At the very beginning of this discussion I said that paths are not consistent across device (dis)connections, like if it was an issue. My initial goal was to setup a reliable identification process of physical devices with the same vid's and pid's, e.g always recognize the one with a red sticker and the one with a green sticker, which I realize is not feasible due to the frequent lack of serial numbers. But it's not a problem anymore now that we're considering a device identification step in the user's workflow. For me the constraint of not touching the physical device connections between initial setup (identification) and end of performance has become acceptable, and @ben-wes I think you nailed the solution in the above quote with a few very clear words. So I second your suggestion : to overcome the scrambled device list issue on macos, we just need to get the list of paths out of [hidraw] (which I think is what you already implemented @Lucarda), and also a way to connect devices by path (e.g an open-path message similar to open-vidpid). How these paths are encoded doesn't sound critical to me as long as the path symbols we get from the outlet can be reused in connection messages.

I still need to double check that the paths don't change between calls to listdevices and are consistent across instances of [hidraw] on macos. It _seems_ to me that the "locations" in [hid] represent the physical USB plugs, I will investigate this tomorrow too. What is exactly "path" supposed to represent ?

**Comment by Lucarda** [comment]
> What is exactly "path" supposed to represent ?

on Windows (this is the problem, see below [*]):

```
path: \\?\HID#VID_13BA&PID_0017&MI_01&Col02#8&22796146&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
```

on Linux:

```
path: /dev/hidraw2
```

on macOS:

do [listdevices( and watch the console.

> and @ben-wes I think you nailed the solution in the above quote with a few very clear words. So I second your suggestion : to overcome the scrambled device list issue on macos, we just need to get the list of paths out of [hidraw] (which I think is what you already implemented @Lucarda), 

no this is not implemented yet. `listvidpid` branch implements something as this:

![hidraw1](https://github.com/user-attachments/assets/ae043cd1-373c-450e-8835-233f587ef707)


I'm hoping this works for you on macOS but if not:

> we just need to get the list of paths out of [hidraw]

> and also a way to connect devices by path (e.g an open-path message similar to open-vidpid). How these paths are encoded doesn't sound critical to me as long as the path symbols we get from the outlet can be reused in connection messages.

this will be implemented in a new `[hidraw/devlist]` object with 4 outlets:

`<symbol :maunufacter> | <symbol :product> | <list float :VID PID>  | <list float :path>`

that one can use to fill a [text] object.

then `[hidraw]` will have a 2th inlet that receives the path as a list of bytes


```
[t b b         ]
|              \
[openpath (    [12 45 78 123 123 .. (
|              /
[hidraw        ]
```

> as long as the path symbols we get from the outlet can be reused in connection messages.

[*] to make it cross platform we need to encode the path as a list of raw bytes.

**Comment by Lucarda** [comment]
@josephlarralde

if you are in a critical mission with deadlines go to the store and buy another model of gamepad. then you get 2 different VID PID and you are done with the red blue sticker.

:)

i'm not saying i will stop development but just saying.

:)

**Comment by ben-wes** [comment]
> then `[hidraw]` will have a 2th inlet that receives the path as a list of bytes

@Lucarda i'm not sure i can follow here - why is that 2nd inlet necessary? couldn't that be an `open-path` message with `A_GIMME` arguments?

and yeah ... in general, this all is adding quite some complexity on the `hidraw` side just because the enumerations of hidapi get scrambled on the same machine with the same connected devices for repeated calls. :/

also seems like quite some overhead to invent a new object for this. but obviously, i see where you're coming from after the previous discussion (adding the path string as a list of ASCII values)!

**Comment by josephlarralde** [comment]
> if you are in a critical mission with deadlines go to the store and buy another model of gamepad. then you get 2 different VID PID and you are done with the red blue sticker.

Well, I get the hint :)
Unfortunately this is not an option, I have to deal with the hardware that is assigned to the project (and anyway the assignation workflow is acceptable). It's a pedagogical project for which there is no budget left at all, and what I'm doing is prioritary to the users but not as much to my supervisors.

> this will be implemented in a new [hidraw/devlist] object with 4 outlets:
>
> `<symbol :maunufacter> | <symbol :product> | <list float :VID PID>  | <list float :path>`
>
> that one can use to fill a [text] object.
>
> then [hidraw] will have a 2th inlet that receives the path as a list of bytes

Sounds good to me. As I'm in a rush to wrap the project up as cleanly as possible to go back to my more prioritary tasks, I will then stick with [hid] for now and watch this repo so that I can substitute my [hid]-based abstractions with [hidraw]-based ones once the new features are ready.

So please take your time to make sure you have the api you really think is best. A list of bytes sounds a bit low-level for me, but I'm totally ok with it (indeed easily retrievable with [text search]) if you think it's the best option. After all we're in pure data and the object name is hidraw so it all makes sense :)

**Comment by josephlarralde** [comment]
@Lucarda I just tested your latest branch.

I can confirm that on macos the paths are consistent across calls to [listdevices<, calls to [list-vidpid<, and also across [hidraw] instances. So far, so good, everything works as expected.

> on macOS:
> 
> do [listdevices( and watch the console.

You are probably already aware of this, but for information on macos the paths all look like : 
* `DevSrvsID:4295324212`
* `DevSrvsID:4295324230`
* `DevSrvsID:4295324267`
* etc ...

I guess your point in favor of a byte list encoding is because the path strings on windows are too long to fit in a symbol ?
It would likely require some extra work, but I think it should be possible to have a dictionary structure to keep a correspondance between the internal representations of the paths and some more digest representations of them (the dictionnary keys) as symbols.

But as I said I'm totally fine using lists of bytes, just preaching for user-friendliness, as long as my suggestion makes sense ...

**Comment by ben-wes** [comment]
> [...] because the path strings on windows are too long to fit in a symbol ?

the problem in this case is not the length, but the special characters in some paths that won't function in messages - see https://github.com/Lucarda/pd-hidraw/issues/5#issuecomment-2552534244

... i admit that i don't completely understand all the options of escaping/unescaping in pd though.

**Comment by Lucarda** [comment]
there are good news here:

- no need to encode path, we can use a symbol.
- there's no need of an extra object as this can be handled with an extra outlet
- since there's no encoded path the 3th outlet can be extended later if more info needs to be added to the list

i'll try to get a new branch ASAP.

**Comment by Lucarda** [comment]
here we got branch `feature/newoutlet`

https://github.com/Lucarda/pd-hidraw/tree/feature/newoutlet

mostly done:


![hidraw2](https://github.com/user-attachments/assets/8342bab7-6b1e-4e08-83a9-9dbd96f24c7a)


@ben-wes and @josephlarralde let me know what you think. if all looks good i'll try to release may be next week. May be tomorrow i'll upgrade hidapi and pd-lib-builder and also setup github actions so that artifacts get built.

**Comment by ben-wes** [comment]
> let me know what you think

tested this quickly here to open a device with `open-path`. works well for me on macos. thanks!

there might obviously be other options to achieve the same goal - but i'm not sure if they are preferable options. so this is good to go from my perspective. some random thoughts i had now (but i think, there's no need to invest that time):
* the additional outlet could be created based on some creation argument (like `-i` for info outlet or so) - which might also silence the console output
* it might be nice to document that for `list-vidpid`, vid and pid also accept values of 0 (which means: no filter for that field)
* the output could also be structured in ways like: `manufacturer <manufacturer_1> <manufacturer_2>, etc.` ... that would make the handling of opening multiple filtered devices with their path a bit simpler. but it would make the vid/pid output a bit ugly (since it would create `vid <vid_1> <vid_2>, pid <pid_1> <pid_2>`). and it would not be compatible with the `text` approach you currently have.
* populating the text could be simplified: 
  ![image](https://github.com/user-attachments/assets/392a3230-bfed-4396-84a3-7e3afb3601ab)


tl;dr: looks good!

**Comment by josephlarralde** [comment]
Thanks ! It also works for me on macos !!!

I'm having a hard time trying to authorize pd to access the devices on ubuntu 22.04 though, but everything seems to work there too.

If I may ask here although it's totally out of the scope of this discussion : do you have by chance an example file to put in /etc/udev/rules.d you would mind sharing ? I tried to create a `70-pd-hidraw-rule.rules` file where I just put the following line : `SUBSYSTEM=="usb", ATTR{bInterfaceClass}=="03", MODE="0666"` as suggested by the google results, and I also tried with `idVendor` and `idProduct` attributes but no luck so far with all the suggested variants I could find ...

Anyway, we can now keep this issue closed :)

**Comment by Lucarda** [comment]
never tried `/etc/udev/rules.d` but this script does what i need. not sure if its permanent. IIRC i had to run it on each boot.

```
#!/bin/bash

sudo chmod 664 /dev/hidraw*
sudo chgrp plugdev /dev/hidraw*
sudo usermod -a -G plugdev $$$
```

replace `$$$` with username.

**Comment by Lucarda** [comment]
on branch `feature/newoutlet` i had

- updated hidapi
- updated pd-lib-builder
- setup github actions for building artifacts


will work on @ben-wes suggestions

- the additional outlet could be created based on some creation argument (like `-i` for info outlet or so) - which might also silence the console output

- it might be nice to document that for `list-vidpid`, vid and pid also accept values of 0 (which means: no filter for that field)

- populating the text could be simplified

**Comment by josephlarralde** [comment]
> never tried /etc/udev/rules.d but this script does what i need. not sure if its permanent. IIRC i had to run it on each boot.

Thanks a lot for this ! (and of course for all the work you did these last days)
I'll get back to my project as soon as the holiday are over.

Best,
Joseph

**Comment by Lucarda** [comment]
on branch `feature/newoutlet` i had implemented

- additional outlet created based on creation argument `-info` which also silences the console output

- documented that for list-vidpid, vid and pid also accept values of 0 (which means: no filter for that field)

- general cleaning of code and improved help patch

@ben-wes if you got time to review you can also get pre-built binaries from https://github.com/Lucarda/pd-hidraw/actions. just grab the latest.

Seems v0.2.0 is almost done.

**Comment by ben-wes** [comment]
> @ben-wes if you got time to review you can also get pre-built binaries from https://github.com/Lucarda/pd-hidraw/actions. just grab the latest.

thanks for those changes - and cool that the artifacts are created here now! macos arm64 version works for me (after the regular permission foo).

the only thing i notice here now (and probably missed last time, sorry): all serial numbers are `no-serial-number` for me on the info outlet - even for devices that seem to have one according to the console log. i'm also not sure if this value for symbolizing a missing serial number is necessary since the other fields seem to work for actual empty values?

see here - where for the first one's product name is missing and the second one's manufacturer name: 
```
Apple  no-serial-number 1452 33028 DevSrvsID:4294968807;
 Keyboard\ Backlight no-serial-number 1452 0 DevSrvsID:4294970219;
```

**Comment by Lucarda** [comment]
thanks for the review.

i changed all not given values to "empty" as i think is better to have a readable list item instead of a non printing character.

can you check the latests https://github.com/Lucarda/pd-hidraw/actions . I want to know if we got issues with the serial number. as in: the serial is reported in the console but not in the 3th outlet?

**Comment by ben-wes** [comment]
> the serial is reported in the console but not in the 3th outlet?

yep. but i see now that it's probably caused by a missing edit after copy/paste here: https://github.com/Lucarda/pd-hidraw/blob/d845e1fd35dad6f1a39f93fb37a1b4c97e64c6a3/pd-hidraw.c#L85

... that should derive the length from `cur_dev->serial_number`.

**Comment by ben-wes** [comment]
(and besides that, the version with `empty` looks good to me!)

**Comment by Lucarda** [comment]
> yep. but i see now that it's probably caused by a missing edit after copy/paste here:

yep, :)

should be fixed in https://github.com/Lucarda/pd-hidraw/actions/runs/12457856942

so v0.2.0 is done?

**Comment by ben-wes** [comment]
> so v0.2.0 is done?

lgtm ... didn't extensively test, but the stuff i reported is certainly fixed! looks like a christmas present. :)

**Comment by Lucarda** [comment]
with some delay:


yesterday i released v0.2.0. waiting for the Windows version to show up on deken and i'll announce it to the list.

**Comment by josephlarralde** [comment]
Yay ! Thanks for the present Santa Lucarda :)

**Comment by josephlarralde** [comment]
Howdy, and happy new year !

If anyone ends up here trying to figure out how to permanently add permissions for hidraw from `/etc/udev/rules.d` on linux, I found the right line to add [here](https://doc.sccode.org/Guides/HID_permissions.html)

TL;DR : it's `KERNEL=="hidraw*", SUBSYSTEM=="hidraw", MODE="0664", GROUP="plugdev"` and you can update the permissions without rebooting with the following commands : 
* `sudo udevadm control --reload-rules`
* `sudo service udev restart`
* `sudo udevadm trigger`

Of course you have to make sure that the current user is part of `plugdev` group fo it to work.

Cheers !
